queue-exp/pifo_srpt.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+# coding: utf-8 -*-
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+#
+# pifo-srpt.py
+
+"""Shortest Remaining Processing Time (SRPT).
+
+This scheduling algorithm is referenced in companion C++ implementation for the
+paper "Programmable packet scheduling at line rate" by Sivaraman, Anirudh, et
+al.
+
+It schedules packets in the order of how much data the flow has left. It assumes
+complete knowledge of the flow length. In the real world, this would either need
+to be estimated or limited to predictable flows.
+"""
+
+__copyright__ = """
+Copyright (c) 2021 Toke Høiland-Jørgensen <toke@toke.dk>
+Copyright (c) 2021 Frey Alfredsson <freysteinn@freysteinn.com>
+"""
+
+__license__ = """
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+from pifo_lib import Packet, Runner, Pifo
+from pifo_lib import SchedulingAlgorithm
+from pifo_lib import FlowTracker
+
+
+class Srpt(SchedulingAlgorithm):
+    """Shortest Remaining Processing Time"""
+
+    def __init__(self):
+        self._pifo = Pifo()
+        self._flow_tracker = FlowTracker()
+
+        self._remains = {}
+
+        # We cheat by accessing the global packet list directly
+        for pkt in pkts:
```

Comment:
I thought of this as the typical example given in operating systems courses where you are supposed to write a nonexistent and optimal scheduling algorithm. I was not too fond of these examples. After all, it is impossible to implement because it needs complete knowledge of each process's remaining time beforehand. Is it different here? Would we reschedule depending on the currently known flow size instead?

Reply:
Ah, now I see what you're trying to do. I think the idea for an actual SRPT algorithm is that the sender will tag each packet with its remaining size in a custom header; that's how they explain it in the Pfabric paper (see the start of §4). This also fits with the "implementation" in the Pifo code, where it just extracts a ranking from the packet...